### ISSUE TITLE
[Phase 2] Feature Engineering Pipeline

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/clinical-risk-stratification-engine#3]
## Summary

Build feature extraction, preprocessing, and selection pipeline with clinical domain constraints. The pipeline must be sklearn-compatible (fit/transform pattern).

**Parent Issue:** #1
**Depends On:** #2 (Phase 1)

## Objectives

- Extract meaningful features from raw EHR data
- Handle missing values without data leakage
- Select features using RFE with clinical constraints

## Files to Create/Modify

| File | Action | Description |
|------|--------|-------------|
| `src/features/extractor.py` | Create | Feature extraction from raw data |
| `src/features/preprocessor.py` | Create | Missing value handling, encoding, scaling |
| `src/features/selector.py` | Create | RFE with clinical constraints |
| `configs/clinical_constraints.yaml` | Create | Must-include/exclude feature lists |
| `tests/test_features.py` | Create | Feature pipeline tests |

### Feature Extractor
- [ ] Extract demographic features
- [ ] Compute vital sign summary statistics (mean, min, max, trend)
- [ ] Compute lab value summary statistics
- [ ] Create binary flags for abnormal values
- [ ] Generate feature names list

### Preprocessor
- [ ] Implement mean imputation strategy
- [ ] Implement median imputation strategy
- [ ] Implement missing indicator strategy
- [ ] One-hot encode categorical variables
- [ ] StandardScaler for numeric features
- [ ] Ensure no data leakage (fit on train, transform on test)

### Feature Selector
- [ ] Implement recursive feature elimination (RFE)
- [ ] Load clinical constraints from YAML
- [ ] Force inclusion of must-have features (e.g., age, vital signs)
- [ ] Force exclusion of prohibited features (e.g., patient_id)
- [ ] Return selected feature mask

### Preprocessor Structure

```python
# src/features/preprocessor.py

from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.impute import SimpleImputer
import pandas as pd
import numpy as np


class EHRPreprocessor(BaseEstimator, TransformerMixin):
    """
    Preprocessor for EHR data that handles missing values,
    encoding, and scaling.

    Learning Context:
    - BaseEstimator and TransformerMixin make this sklearn-compatible
    - fit() learns parameters from training data
    - transform() applies those parameters to new data
    - This prevents data leakage (using test info during training)
    """

    def __init__(self, impute_strategy='median'):
        self.impute_strategy = impute_strategy
        self.numeric_features = None
        self.categorical_features = None
        self.preprocessor = None

    def fit(self, X: pd.DataFrame, y=None):
        """Learn preprocessing parameters from training data."""
        self.numeric_features = X.select_dtypes(include=[np.number]).columns.tolist()
        self.categorical_features = X.select_dtypes(include=['object', 'category']).columns.tolist()

        # Remove patient_id if present
        self.numeric_features = [f for f in self.numeric_features if f != 'patient_id']
        self.categorical_features = [f for f in self.categorical_features if f != 'patient_id']

        numeric_pipeline = Pipeline([
            ('imputer', SimpleImputer(strategy=self.impute_strategy)),
            ('scaler', StandardScaler())
        ])

        categorical_pipeline = Pipeline([
            ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
            ('encoder', OneHotEncoder(drop='first', sparse_output=False, handle_unknown='ignore'))
        ])

        self.preprocessor = ColumnTransformer([
            ('numeric', numeric_pipeline, self.numeric_features),
            ('categorical', categorical_pipeline, self.categorical_features)
        ])

        self.preprocessor.fit(X)
        return self

    def transform(self, X: pd.DataFrame) -> np.ndarray:
        """Apply preprocessing to new data."""
        return self.preprocessor.transform(X)

    def get_feature_names_out(self):
        """Get output feature names after transformation."""
        return self.preprocessor.get_feature_names_out()
```

### Clinical Constraints Configuration

```yaml
# configs/clinical_constraints.yaml

must_include:
  - age
  - heart_rate
  - systolic_bp
  - diastolic_bp
  - temperature
  - respiratory_rate

must_exclude:
  - patient_id
  - admission_date
  - discharge_date

min_features: 10
max_features: 50
```

### Feature Selector Structure

```python
# src/features/selector.py

from sklearn.feature_selection import RFE
from sklearn.ensemble import RandomForestClassifier
import yaml
import numpy as np


class ClinicalFeatureSelector:
    """
    Feature selector using RFE with clinical domain constraints.

    Ensures clinically important features are always included
    and prohibited features are always excluded.
    """

    def __init__(self, config_path='configs/clinical_constraints.yaml'):
        with open(config_path) as f:
            self.config = yaml.safe_load(f)

        self.must_include = set(self.config.get('must_include', []))
        self.must_exclude = set(self.config.get('must_exclude', []))
        self.n_features = self.config.get('max_features', 30)
        self.selector = None
        self.feature_names = None
        self.selected_mask = None

    def fit(self, X: np.ndarray, y: np.ndarray, feature_names: list):
        """Fit selector with clinical constraints."""
        self.feature_names = feature_names

        # Create mask for must-include features
        must_include_mask = np.array([
            f in self.must_include for f in feature_names
        ])

        # Run RFE on remaining features
        base_estimator = RandomForestClassifier(n_estimators=50, random_state=42)
        self.selector = RFE(base_estimator, n_features_to_select=self.n_features)
        self.selector.fit(X, y)

        # Combine: must-include OR selected by RFE
        self.selected_mask = must_include_mask | self.selector.support_

        return self

    def transform(self, X: np.ndarray) -> np.ndarray:
        """Select features based on fitted mask."""
        return X[:, self.selected_mask]

    def get_selected_features(self) -> list:
        """Get names of selected features."""
        return [f for f, m in zip(self.feature_names, self.selected_mask) if m]
```

## Verification

```bash
# Run tests
pytest tests/test_features.py -v

# Quick verification
python -c "
from src.features.preprocessor import EHRPreprocessor
from src.data.synthetic import generate_synthetic_ehr

df = generate_synthetic_ehr(1000)
X = df.drop(columns=['readmission', 'sepsis', 'mortality'])

prep = EHRPreprocessor()
X_transformed = prep.fit_transform(X)
print(f'Original shape: {X.shape}')
print(f'Transformed shape: {X_transformed.shape}')
print(f'Feature names: {prep.get_feature_names_out()[:5]}...')
"
```

## Definition of Done

- [ ] Preprocessor handles missing values without data leakage
- [ ] Preprocessor encodes categoricals and scales numerics
- [ ] Feature extractor produces 50+ features from raw EHR
- [ ] Selector respects must-include constraints
- [ ] Pipeline is sklearn-compatible (fit/transform)
- [ ] All unit tests pass
- [ ] Code has docstrings

## Technical Notes

**Learning Context:**
- Understand sklearn's fit/transform pattern (critical for ML pipelines)
- Learn about data leakage and why it's dangerous
- Study one-hot encoding and when to use it

**Potential Challenges:**
- Getting feature names after ColumnTransformer transformations
- Handling high-cardinality categoricals (ICD codes have 10,000+ values)
